### ISSUE TITLE
Client dropped expired request in multithreading environment

### ISSUE THREAD

**Issue by zy1994-lab** [google/tarpc#258]
Hi,

I am running a RPC server in many machines and accessing it using many clients each has 3 threads from another machines. However, I got the following error:

```
thread 'worker thread 2' panicked at 'RPC error:Custom { kind: TimedOut, error: "Client dropped expired request." }',
thread 'worker thread 0' panicked at 'RPC error:Custom { kind: TimedOut, error: "Client dropped expired request." }', 
thread 'worker thread 1' panicked at 'RPC error:Custom { kind: TimedOut, error: "Client dropped expired request." }', 
```

I put all clients to multiple server in a `vec` and create the clients in each thread.

```rust
pub struct GraphClient {
    server_addrs: Vec<SocketAddr>,
    runtime: RefCell<tokio::runtime::current_thread::Runtime>,
    clients: Vec<Option<RefCell<GraphRPCClient>>>,
}

impl GraphClient {
    fn create_clients(&mut self) {
        for (i, addr) in self.server_addrs.iter().enumerate() {
                let transport = self
                    .runtime
                    .borrow_mut()
                    .block_on(async move {
                        info!("Connecting to {:?}", addr);
                        bincode_transport::connect(&addr).await
                    })
                    .unwrap_or_else(|e| panic!("Fail to connect to {:}: {:}", addr, e));

                let NewClient { client, dispatch } =
                    GraphRPCClient::new(client::Config::default(), transport);

                self.runtime.borrow_mut().spawn(async move {
                    if let Err(e) = dispatch.await {
                        panic!("Error while running client dispatch: {:?}", e)
                    }
                });

                Some(RefCell::new(client))
            };

            self.clients.push(client);
        }
}
```

How should I config `max_channels_per_key`, `buffer_unordered` or `max_concurrent_requests_per_channel` for the server?

**Comment by zhengyi-yang** [comment]
Actually I connected to multiple RPC server in each thread. So I put all clients in a `vec`. The code looks like the follow:

```rust
pub struct GraphClient {
    server_addrs: Vec<SocketAddr>,
    runtime: RefCell<tokio::runtime::current_thread::Runtime>,
    clients: Vec<Option<RefCell<GraphRPCClient>>>,
}

impl GraphClient {
    fn create_clients(&mut self) {
        for (i, addr) in self.server_addrs.iter().enumerate() {
                let transport = self
                    .runtime
                    .borrow_mut()
                    .block_on(async move {
                        info!("Connecting to {:?}", addr);
                        bincode_transport::connect(&addr).await
                    })
                    .unwrap_or_else(|e| panic!("Fail to connect to {:}: {:}", addr, e));

                let NewClient { client, dispatch } =
                    GraphRPCClient::new(client::Config::default(), transport);

                self.runtime.borrow_mut().spawn(async move {
                    if let Err(e) = dispatch.await {
                        panic!("Error while running client dispatch: {:?}", e)
                    }
                });

                Some(RefCell::new(client))
            };

            self.clients.push(client);
        }
}
```

**Comment by vorot93** [comment]
I believe this is a regression caused by #240. We were bitten very hard by this but never had the time to investigate and thus use an older revision of tarpc at the moment.

**Comment by zy1994-lab** [comment]
@vorot93 Thanks for your reply. I managed to get it work by setting `max_concurrent_requests_per_channel` to a higher value.

**Comment by tikue** [comment]
Sorry for the pain caused by this! I know it's hard to follow all changes being made on master. I let this particular PR sit for like a month in the hopes of gaining more visibility on it.

Do you have any suggestions for how to avoid issues like this? The only one I have is to make more frequent releases with detailed release notes, so that people are more likely to be aware of the changes.

**Comment by tikue** [comment]
@vorot93 do you believe there is a bug in the code, or do you just mean it's more difficult to get it configured appropriately? Interested to hear any thoughts you have.

**Comment by tikue** [comment]
@zy1994-lab if the requests are expiring, that means the server isn't even seeing them. `max_concurrent_requests_per_channel` just means the server won't throttle the client until the value specified. But throttling doesn't lead to expired requests. The client would instead receive an error response from the server.

I suspect you're not executing either the server channels or each channel's requests concurrently. Can you show me the server code?

**Comment by vorot93** [comment]
```
#[macro_export]
macro_rules! listen {
    ($addr:expr, $service_factory:expr) => {
        tarpc_json_transport::listen(&$addr)?
            .filter_map(|r| $crate::rpc::futures::future::ready(r.ok()))
            .map($crate::rpc::tarpc::server::BaseChannel::with_defaults)
            .map(move |channel| channel.respond_with(($service_factory)()))
            .for_each(|_| $crate::rpc::futures::future::ready(()))
    };
}
```

This is how we tried to use the new tarpc server.

**Comment by tikue** [comment]
@vorot93 I can't look closely right now, but in that macro, the future returned by `channel.respond_with` appears to be dropped on the floor in the `for_each`. I'd expect that to cause clients to be immediately disconnected.

**Comment by zy1994-lab** [comment]
@tikue Thanks Tim, I used the filling server code:

```rust
#[derive(Clone)]
pub struct Server {
    graph: Arc<Vec<u32>>,
}

impl GraphServer {
    pub async fn run(
        self,
        port: u16,
        machines: usize,
        workers: usize,
    ) -> io::Result<()> {
        let server_addr = ([0, 0, 0, 0], port).into();

        let transport = bincode_transport::listen(&server_addr)?;

        transport
            .filter_map(|r| future::ready(r.ok()))
            .map(server::BaseChannel::with_defaults)
            .max_channels_per_key(workers as u32, |t| {
                t.as_ref().peer_addr().unwrap().ip()
            })
            .max_concurrent_requests_per_channel(workers)
            .map(|channel| {
                let server = self.clone();

                channel.respond_with(server.serve()).execute()
            })
            .buffer_unordered(workers*machines)
            .for_each(|_| async {})
            .await;

        Ok(())
    }
```

Here, `machines` is the number of machines in my cluster, each running a RPC server, 'workers' is the number of clients running in each machine connecting to all other servers in the same cluster. I don't understand why, but config as above work. Can you get me some advise?

**Comment by tikue** [comment]
@zy1994-lab my guess is it's the `.buffer_unordered(workers*machines)` that fixed things for you. Without that, you'd only run a maximum of one channel at a time, so only one client would ever be able to have requests fulfilled.

BTW, you may also consider spawning the future returned by `channel.respond_with(server.serve()).execute()` -- otherwise, all of your server's deserialization and serialization will happen on one thread.

**Comment by zhengyi-yang** [comment]
@tikue Thanks Tim. I have changed the code to 

```rust
.map(|channel| {
                let server = self.clone();

                let f = async {
                    runtime.spawn(channel.respond_with(server.serve()).execute());
                };

                f
            })
```

Is that correct?

**Comment by tikue** [comment]
@zhengyi-yang that gets rid of the limit on channels entirely, in addition to running each channel concurrently. That may or may not be what you want. If you want to run your channels in parallel but still limit the # of connections, you could do this:

```rust
.map(|channel| {
    let server = self.clone();
    let (tx, rx) = oneshot::channel();
    runtime.spawn(async move {
        channel
            .respond_with(server.serve())
            .execute()
            .await;
        tx.send(());
    });
    rx
})
.buffer_unordered(workers*machines)
```

This will have the main task handle thin handles to the channels so that it can still limit the number of channels running. (Aside: I thought tokio had a `spawn_with_handle` function that made this pattern easier to do, but perhaps it's not been added yet)

**Comment by zy1994-lab** [comment]
@tikue Thanks. That's vey helpful!

**Comment by zy1994-lab** [comment]
Hi @tikue , I'm still trying to optimise the RPC performance in my program. In the case of multiple threads in a machine connecting to one RPC server, there are 2 ways of doing this. One is to create a channel and a client for each thread, the other is to run only one client and cloning it across all threads.  When I looked inside the source code, I saw part of the Client is warped in `Mutex`, but I didn't figure out how cloning the client across thread will work. Does that mean all RPC happens in one thread? What's the best practice to sharing clients among threads?

PS: I'm using a `tokio::runtime::Runtime` for the clients when cloning the client, and `tokio::runtime::current_thread::Runtime` in each worker thread when creating client independently.

Thanks so much.
